### ISSUE TITLE
Error in Numba-compiled DPOSV call

### ISSUE THREAD

**Issue by mdekstrand** [numba/numba#6556]
- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

I have a Numba-compiled function that wraps the BLAS `dposv` function. This function was working correctly up through Numba 0.51. With Numba 0.52, it fails with a DPOSV error. The following code demonstrates the problem.

- Under Numba 0.51, it runs correctly
- Under Numba 0.52 (on Windows, installed from conda-forge), the assertion fails (indicating a DPOSV argument error) and the BLAS implementation (tested with both MKL and OpenBLAS) complains about an invalid argument

As near as I can tell, Numba 0.52 changed something about how this function compiles so it no longer functions. This function is a part of [LensKit](https://github.com/lenskit/lkpy), and it can be run with LensKit's tests as well:

    python -m pytest -k dposv

```python
import numpy as np
import numba as n
import cffi
from numba.extending import get_cython_function_address

__ffi = cffi.FFI()

__uplo_U = np.array([ord('U')], dtype=np.int8)
__uplo_L = np.array([ord('L')], dtype=np.int8)
__dposv = __ffi.cast("void (*) (char*, int*, int*, double*, int*, double*, int*, int*)",
                     get_cython_function_address("scipy.linalg.cython_lapack", "dposv"))


@n.njit(n.intc(n.float64[:, ::1], n.float64[::1], n.boolean))
def _dposv(A, b, lower):
    if A.shape[0] != A.shape[1]:
        return -11
    if A.shape[0] != b.shape[0]:
        return -12

    # dposv uses Fortran-layout arrays. Because we use C-layout arrays, we will
    # invert the meaning of 'lower' and 'trans', and the function will work fine.
    # We also need to swap index orders
    uplo = __uplo_U if lower else __uplo_L
    n_p = __ffi.from_buffer(np.array([A.shape[0]], dtype=np.intc))
    nrhs_p = __ffi.from_buffer(np.ones(1, dtype=np.intc))
    info = np.zeros(1, dtype=np.intc)
    info_p = __ffi.from_buffer(info)

    __dposv(__ffi.from_buffer(uplo), n_p, nrhs_p,
            __ffi.from_buffer(A), n_p,
            __ffi.from_buffer(b), n_p,
            info_p)

    return info[0]

A = np.array([[106014.2749504 ,  53086.26042087],
       [ 53086.26042087, 252421.30232019]])
b = np.array([ 4875.45574154, -4608.45667769])
rv = _dposv(A, b, True)
assert rv >= 0
```

**Comment by mdekstrand** [comment]
I have used git-bisect to hunt down a source for this bug, and it seems to be introduced in Numba revision 4799601f909afb725f20812c948f2c53bf096998.

**Comment by mdekstrand** [comment]
I have tried with Numba 0.52 with `NUMBA_LLVM_REFPRUNE_PASS` set to `0`, and still see the error, so I can't be 100% confident refprune is to blame, but the change that turned it on breaks my function.

**Comment by esc** [comment]
@mdekstrand thank you for reporting this. I can confirm I am able to reproduce this.  The reproducer was missing the following imports:

```
import numba as n
import numpy as np
```

It seems to work fine on 0.51.2 and fails on current master (`0.53.0dev0-335-g484925571c`) with:

```
 zsh» python issue6556.py

Intel MKL ERROR: Parameter 5 was incorrect on entry to DPOSV .
Traceback (most recent call last):
  File "issue6556.py", line 41, in <module>
    assert rv >= 0
AssertionError
```

**Comment by esc** [comment]
I can also confirm that bisect will lead to `4799601f909afb725f20812c948f2c53bf096998`.

**Comment by stuartarchibald** [comment]
I think the issue here is that the original code worked by "accident" and the more aggressive reference pruner/newly tuned optimiser exposes this.

Essentially, after the LAPACK `DPOSV` call, there's nothing keeping the references to the reference counted variables (arrays in this case) "alive". The error coming from MKL is likely due to it reading nonsense memory as a result of a free occurring as a result of the refcount reaching zero before the LAPACK call. The fix is thankfully easy, and a technique Numba uses internally for it's `np.linalg` support, which is the use of a dummy function that does nothing apart from act as a sink for reference counted variables such that they are kept alive.

e.g.
```python
import numpy as np
import numba as n
import cffi
from numba.extending import get_cython_function_address

__ffi = cffi.FFI()

__uplo_U = np.array([ord('U')], dtype=np.int8)
__uplo_L = np.array([ord('L')], dtype=np.int8)
__dposv = __ffi.cast("void (*) (char*, int*, int*, double*, int*, double*, int*, int*)",
                     get_cython_function_address("scipy.linalg.cython_lapack", "dposv"))

@n.njit
def ref_sink(*args):
    return args

@n.njit(n.intc(n.float64[:, ::1], n.float64[::1], n.boolean))
def _dposv(A, b, lower):
    if A.shape[0] != A.shape[1]:
        return -11
    if A.shape[0] != b.shape[0]:
        return -12

    # dposv uses Fortran-layout arrays. Because we use C-layout arrays, we will
    # invert the meaning of 'lower' and 'trans', and the function will work fine.
    # We also need to swap index orders
    uplo = __uplo_U if lower else __uplo_L
    n_p = __ffi.from_buffer(np.array([A.shape[0]], dtype=np.intc))
    nrhs_p = __ffi.from_buffer(np.ones(1, dtype=np.intc))
    info = np.zeros(1, dtype=np.intc)
    info_p = __ffi.from_buffer(info)

    __dposv(__ffi.from_buffer(uplo), n_p, nrhs_p,
            __ffi.from_buffer(A), n_p,
            __ffi.from_buffer(b), n_p,
            info_p)

    ref_sink(__uplo_U, n_p, nrhs_p, info, info_p)

    return info[0]

A = np.array([[106014.2749504 ,  53086.26042087],
       [ 53086.26042087, 252421.30232019]])
b = np.array([ 4875.45574154, -4608.45667769])
rv = _dposv(A, b, True)

_dposv.inspect_cfg(_dposv.signatures[0]).display(view=True, filename='with_sink',
                                                 format='svg')
assert rv >= 0
```
The impact of this is as follows, this is the original CFG, note how the decrefs occur before the `%.999 = tail call i8* %.9054` (this is the call to LAPACK):
![no_sink](https://user-images.githubusercontent.com/907717/101472864-11400d80-3941-11eb-8c1c-b703e364fe60.png)

Whereas in case of using the sink function, the CFG has all the decrefs at the end after the LAPACK call:
![with_sink](https://user-images.githubusercontent.com/907717/101472917-1dc46600-3941-11eb-9ddc-af40d110e465.png)



Hope this helps?

**Comment by mdekstrand** [comment]
Yes, I can confirm that this fixes the problem in the demo and in my primary code base.

It's rather unintuitive, and seems to break the general expectation of Python reference counting (as Python does not, to my knowledge, do any reachability analysis to pre-emptively release references), but the reference sink fixes the problem.

**Comment by stuartarchibald** [comment]
> Yes, I can confirm that this fixes the problem in the demo and in my primary code base.

Great, thanks for confirming.

> It's rather unintuitive, and seems to break the general expectation of Python reference counting (as Python does not, to my knowledge, do any reachability analysis to pre-emptively release references), but the reference sink fixes the problem.

Numba implements its own internal reference counting system, https://numba.readthedocs.io/en/stable/developer/numba-runtime.html, it cooperates with the CPython interpreter but is not the same thing. Fortunately it remains almost entirely hidden to users, until there's cases like this. I imagine with some quite involved work this sort of thing can be spotted and fixed automatically, but I suspect there's going to be cases where it cannot be.

**Comment by stuartarchibald** [comment]
Changed title to hopefully make it more easily found by others with a similar issue.

**Comment by mdekstrand** [comment]
Thanks. Two outstanding questions:

1. Is this edge case documented somewhere discoverable? I'm not even sure what the right place and way to document it is, because the signals that you need to apply this technique are so opaque. Does it occur in situations that don't involve cffi/ctypes?
2. Is this at risk of breaking again, e.g. the optimizer inlining the `ref_sink` function and then the refprune pass moving the decrefs back up again?

**Comment by stuartarchibald** [comment]
> Thanks. Two outstanding questions:
> 
>     1. Is this edge case documented somewhere discoverable? I'm not even sure what the right place and way to document it is, because the signals that you need to apply this technique are so opaque. Does it occur in situations that don't involve cffi/ctypes?

Don't think this is documented, it's pretty rare and as noted, opaque. It can occur anywhere where there's a function consuming pointer-y things from reference counted objects.

>     2. Is this at risk of breaking again, e.g. the optimizer inlining the `ref_sink` function and then the refprune pass moving the decrefs back up again?

It's unlikely, the inliner is already inlining the sink function. I spoke with @sklam a bit about various options including supplying helper functions, doing code gen based on analysis etc. But the most likely "fix" is to just push all decrefs to the end of the basic block, this extends the lifetime of objects slightly but should prevent situations like this.

**Comment by stuartarchibald** [comment]
Reclassified as a miscompile.

**Comment by stuartarchibald** [comment]
Potential fix in https://github.com/numba/numba/pull/6623
